src/flash.rs

```diff
@@ -101,6 +114,111 @@ impl<'a> FlashWriter<'a> {
         }
     }
 
+    fn unlock_opt(&mut self) -> Result<()> {
+        // First we must unlock the FPEC
+        self.unlock()?;
+
+        // Wait for any ongoing operations
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        // NOTE(unsafe) write Keys to the OPTKEYR register. This is safe because the
+        // only side effect of these writes is to unlock the option bytes
+        // register, which is the intent of this function. Do not rearrange the
+        // order of these writes.
+        self.flash
+            .optkeyr
+            .optkeyr()
+            .write(|w| unsafe { w.optkey().bits(KEY1) });
+
+        self.flash
+            .optkeyr
+            .optkeyr()
+            .write(|w| unsafe { w.optkey().bits(KEY2) });
+
+        // Verify success
+        if self.flash.cr.cr().read().optwre().bit_is_set() {
+            Ok(())
+        } else {
+            Err(Error::UnlockOptError)
+        }
+    }
+
+    fn lock_opt(&mut self) -> Result<()> {
+        // Reset the OPTWRE bit before relocking flash generally
+        self.flash.cr.cr().write(|w| w.optwre().clear_bit());
+        self.lock()
+    }
+
+    /// Erases the Option Bytes. Requires Option Bytes registers to be unlocked.
+    fn erase_opt(&mut self) -> Result<()> {
+        // Ensure OPTWRE is set
+        if self.flash.cr.cr().read().optwre().bit_is_set() {
+            // Erase the option bytes and wait
+            self.flash.cr.cr().modify(|_, w| w.opter().set_bit());
+            self.flash.cr.cr().modify(|_, w| w.strt().set_bit());
+            while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+            // Clear EOP bit if set
+            if self.flash.sr.sr().read().eop().bit_is_set() {
+                self.flash.sr.sr().modify(|_, w| w.eop().clear_bit());
+            }
+            // Clear the OPTER bit
+            self.flash.cr.cr().modify(|_, w| w.opter().clear_bit());
+            Ok(())
+        } else {
+            Err(Error::UnlockOptError)
+        }
+    }
+
+    fn set_rdp(&mut self, val: u8) -> Result<()> {
+        // First we need to unlock the Option Byte programming
+        self.unlock_opt()?;
+
+        // Wait for operation to finish
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        // First we must erase the Option Bytes
+        self.erase_opt()?;
```

Comment:
Taking a second look -- this needs to retain the `USER` byte value and others. Maybe we can get away without erasing them entirely? The documentation is a little ambiguous.

Reply:
Update on this:

Writing the RDP value will fail if we haven't erased the Option Bytes first. I've tried saving off the upper half-word and writing it with the new RDP value, but I can't get it to accept it.  Will have to investigate more.